### PR TITLE
whole_list_condition_test flakiness fix

### PR DESCRIPTION
**Pull request by knifewine**
write and read back at more appropriate levels
write at cl.all and read back at cl.one

it's possible that this test was failing previously
when a coordinator (and selfsame replica) would defer
to another node at cl.one, allowing a stale data read
and therefore an assertion failure

### CODE REVIEW

**Comment by knifewine** [comment]
@mambocab to review.

**Comment by knifewine** [comment]
I also snuck in some string formatting changes for readability.

I ran a slightly modified version of the test locally (with a loop going through thousands of repetitions) and it was able to pass without assertion errors.

File: upgrade_tests/cql_tests.py [hunk 1]
```diff
@@ -4267,6 +4267,8 @@ def whole_list_conditional_test(self):
                 l frozen<list<text>>
             )""")
 
+        cl = getattr(self, 'CL', ConsistencyLevel.ONE)
```
Comment (mambocab) [inline upgrade_tests/cql_tests.py]: `self.CL` should always be set; there's a default set on the `UpgradeBase` class:

https://github.com/riptano/cassandra-dtest/blob/master/upgrade_tests/upgrade_base.py#L106

**Comment by mambocab** [comment]
I think we want to run this by @thobbs or @pcmanus -- I thought we wrote and read at `ONE` to ensure we only read from a single node, so maybe the solution here is to restrict the query by some other means.

File: upgrade_tests/cql_tests.py (hunk 1 above)
Comment (knifewine) [inline upgrade_tests/cql_tests.py]: ah, I thought it could be none. pushing a change to that momentarily.

**Comment by knifewine** [comment]
based my a conversation yesterday I'm not certain we can prevent a coordinator who is a replica from going out to another replica in rare cases. 
read_from_others

**Comment by mambocab** [comment]
Hmm, sure. I guess I don't understand the intent of using an exclusive connection, then. If we don't know that the query is going to be executed on particular nodes, then what are we testing with this setup?

**Comment by ptnapoleon** [comment]
In the general case, exclusive guarantees the node will be coordinator, not handle the queries. You probably know that, but worth recording here.

**Comment by ptnapoleon** [comment]
Is the exclusive connection in the `prepare()`? I don't see it in the test proper.

**Comment by knifewine** [comment]
In this case the prepare connection isn't exclusive, but it's not actually
used for the majority of the test.

**Comment by ptnapoleon** [comment]
Oh, it's coming from `do_upgrade` then?

**Comment by knifewine** [comment]
yep

**Comment by knifewine** [comment]
I didn't write this test, so I'm less clear on the unstated goals, but the impression I had was just that we're checking the conditional application of updates. I don't think we need a guarantee of exclusivity for that case anyway. If the test is not valid without an exclusivity guarantee, we can retire it as well.

**Comment by ptnapoleon** [comment]
Okay. The point of the exclusivity is to ensure that in the mixed version cluster we set up, that a node from each version is tested as coordinator, to ensure that queries work in both directions, because bugs won't be symmetrical. 

As far as the conditional application of updates, I 100% think raising the CL is fine, based on my understanding of the issue.

**Comment by knifewine** [comment]
We could keep the low CL and introduce repairs as part of the test as well. not sure if that's the right approach though.

**Comment by ptnapoleon** [comment]
-1 on low CL and repairing.

**Comment by mambocab** [comment]
> ensure that in the mixed version cluster we set up, that a node from each version is tested as coordinator

Seems odd to me that, with RF=ALL, the coordinator would query another node on a read with CL=ONE. As long as that's not a C* bug, I think raising the CL for writes should be fine.

I feel like we should file a ticket to write a different value with each of the updates in this test. Am I off-base?

**Comment by knifewine** [comment]
> Seems odd to me that, with RF=ALL, the coordinator would query another node on a read with CL=ONE. As long as that's not a C* bug, I think raising the CL for writes should be fine

the all before wasn't being respected because assert_one, et al, create statements and apply a default consistency level of one to that statement, so this was superceding what was set on the session.

> I feel like we should file a ticket to write a different value with each of the updates in this test. Am I off-base?

seems like a good idea to me

**Comment by thobbs** [comment]
It sounds like you guys sorted everything out, but just to confirm, the test is primarily concerned with ensuring the conditional updates apply correctly.  Specifically, making sure that all of the operators work correctly with different values and nulls.

> I feel like we should file a ticket to write a different value with each of the updates in this test. Am I off-base?

It would be nice to do this to ensure that all of the paxos writes are happening correctly.  That's not the original purpose of the test, but I think it's a good extension, and especially makes sense for an upgrade test.

**Comment by ptnapoleon** [comment]
+1 to this code, and also to filing a ticket to extend as described by Jim and Tyler

**Comment by mambocab** [comment]
@ptnapoleon I couldn't find the ticket -- could you assign me? i have Ideas.

**Pull request merged**